storage-proofs/core/src/drgraph.rs

```diff
@@ -157,10 +159,9 @@ impl<H: Hasher> Graph<H> for BucketGraph<H> {
                 let metagraph_node = node as u64 * m_prime as u64;
                 let n_buckets = (metagraph_node as f64).log2().ceil() as u64;
 
-                let (predecessor_index, other_drg_parents) = if self.is_legacy {
-                    (m_prime, &mut parents[..])
-                } else {
-                    (0, &mut parents[1..])
+                let (predecessor_index, other_drg_parents) = match self.api_version {
+                    ApiVersion::V1_0_0 => (m_prime, &mut parents[..]),
+                    ApiVersion::V1_1_0 => (0, &mut parents[1..]),
```

Comment:
Here's an idea. What if instead of doing this kind of ad hoc inline matching, we *only* enact conditional code  by calling methods of the `ApiVersion`? The downside: loss of local context in the implementations. The upside: it would be very easy to maintain and inspect an exhaustive set of differences between versions. This example would be something like (untested):
```
trait DRGParentOrdering {
    …
}

impl DRGParentOrdering for ApiVersion::1_0_0 {
    fn parent_ordering(&self, m_prime: usize, parents: &mut[T]) -> (usize, &mut[T]) {
        (m_prime, parents)
    }

impl DRGParentOrdering for ApiVersion::1_1_0 {
    fn parent_ordering(&self, m_prime: usize, parents: &mut[T]) -> (usize, &mut[T]) {
        (0, parents[1..])
    }
}
```

… and then:
```
let (predecessor_index, other_drg_parents) = self.api_version.parent_ordering(m_prime, parents);
```

Add appropriate comments, and this becomes a very clear and self-documenting record of exactly what has changed between versions, as well as for each atomic patch's dependencies.

In addition to the self-documenting property, this would help enforce a discipline to combat the almost inevitable spaghetti-like nature of changes likely to accumulate. By establishing and enforcing such a pattern, you would disable and disarm the temptation to just make the implementation a little more complicated, a little more clever — a complexity which might accumulate in the name of simplicity and be difficult to untangle later. Such accumulations and potential multiplicative complexity are one of the things that worries me most about this path (although I understand why its _general_ shape is probably the right one now).

This is one reason I think the right discipline and a strong set of rules (even if they have to be amended in accordance with need over time) will go a long way toward protection. If something is too difficult to do within the trait-patch method, it might be a *good idea* to rethink it. Admittedly, this proposal has the disadvantage of moving the actual logic out of line. I think this could be solved with a little tooling. One idea: a macro which inlines the values for the *current — i.e. correct* version and defers to to trait implementations for the others.

```
let (predecessor_index, other_drg_parents) = with_version_or!(DRGParentOrdering, (0, &mut [parents[..]));
```

This assumes there *is* a concept of a 'current version', and that it is tracked in the code, but I think that may be a good/safe assumption.

Again, I realize this is more work, and introduces a certain amount of awkwardness. At the same time, I think the rigidity can be protective by:

- Making it *harder* to do ad hoc things or to introduce unexpected variations or multiplicative complexity into the code.
- Isolating and forcing centralized documentation of all version deltas.
- Forcing a higher level of discipline around the form of allowable version-based changes. Nothing prevents gradually extending the complexity of what is allowed, but the *need* to do that or else view the result of shortcutting it as at least technical debt, might force more deliberate attention around each such change introduced and how they interact.

I realize I this all comes across as very grave, but as we have discussed before, I think the approach we are taking here (however necessary) skirts a dangerous territory — and it's well worth extra forethought and extra discipline to ensure we stay on the right side of the line. Avoiding the problems is easy *now*, at the first very simple change. It's later, as they stack up, grow in complexity, and need to survive refactors, etc. that it might become hard to deal with. Whatever policies we can put in place to prevent that will be valuable. Enshrining those in code makes it just that much harder to break or change them in the heat of the moment.

Reply:
It's nice for simple cases (like the above), but won't work when parameters or return values change.  Each behavior change will have to be considered differently at some point.

Reply:
I like the idea, but I fear it is not quite practical, as we end up with circular imports. `ApiVersion` is defined in `storage-proofs-core`, but if we were to push all these details onto it, we would have to use types in core from `storage-proofs-porep` and `storage-proofs-post`, resulting in a circle.

One idea to help reduce the random places with api versions matching, is to at least always put the two different versions into a function call, named with a suffix of `_v_1_0_0` or simlar, for example

```rust
impl Bar {
  fn foo(&self, version: ApiVersion) {
    match version {
      ApiVersion::V1_0_0 => self.foo_v1_0_0(),
      ApiVersion::V1_1_0 => self.foo_v1_1_0(),
    }
  }

  fn foo_v1_0_0(&self) {
    todo!()
  }
  fn foo_v1_1_0(&self) {
    todo!()
  }
}

Reply:
@dignifiedquire This is also a nice re-factor for the simple case, but breaks down when args or return types change (at least the dispatch `foo` above.  It could work if there was no dispatch and just separate versioned foo methods -- but that pushes the dispatch (i.e. match logic) up further if the return type changes, which is exactly what's trying to be avoided, no?).  The upshot of these discussions are that they are re-factors, which can be done at any time (and probably will no matter what we choose here), independent of the fact that we've settled on a logic switching mechanism for upgrades.

For this first cut, imposing a formality around something that's known to not work in all cases seems arbitrary, but if it's the will of the team, we can.

Reply:
I agree we don’t need to figure this out before merging

Reply:
Don't let it hold you up. I more wanted to inject some ideas into the mix for as the ongoing refactors @cryptonemo predicts happen. Changes to the API signature are obviously a larger problem that will need to be considered (perhaps soon). That may be somewhat orthogonal.

This is all just continued brainstorming and not meant to be blocking — just as feed for ongoing design and consideration:

@dignifiedquire Just continuing to brainstorm: you could implement the traits in the most local scope they are needed. That would be less centralized, but still easy to inspect and would capture information of itself.